File: apple/RemoteDisplay/Keyboard/KeyboardController.swift
//
//  KeyboardController.swift
//  RemoteDisplay
//
//  Created by Júlio Flores on 01/09/2026.
//

extension Keyboard {
	@MainActor
	final class Controller {
		private var subscriptions: Set<AsyncStream<Keyboard.Action>.Continuation> = []
		private func removeSubscription(_ subscription: AsyncStream<Keyboard.Action>.Continuation) {
			subscriptions.remove(subscription)
		}

		deinit {
			for subscription in self.subscriptions {
				subscription.finish()
			}
		}
	}
}

extension Keyboard.Controller: Keyboard.Receiver {
	func insertText(_ text: String) {
		guard !text.isEmpty else {
			return
		}
		for subscription in self.subscriptions {
			subscription.yield(.text(text))
		}
	}

	func deleteBackward() {
		for subscription in self.subscriptions {
			subscription.yield(.backspace)
		}
	}
}

extension Keyboard.Controller: Keyboard.Forwarder {
	var keyboardAction: AsyncStream<Keyboard.Action> {
		AsyncStream<Keyboard.Action> { continuation in
			self.subscriptions.insert(continuation)
			continuation.onTermination = { _ in
				Task {
					await self.removeSubscription(continuation)
				}
			}
		}
	}

	func endActionStream() {
		for subscription in self.subscriptions {
			subscription.finish()
		}
		subscriptions.removeAll()
	}
}

File: android/app/src/main/java/dk/cipher/remotedisplay/keyboard/KeyboardController.kt
package dk.cipher.remotedisplay.keyboard

import dk.cipher.remotedisplay.keyboard.Keyboard.Action

class KeyboardController : Keyboard.Controller {
    private var service: Keyboard.Service? = null
    private var subscriptions = mutableSetOf<Keyboard.Client>()
    private var isShowingKeyboard = false
    private var _canShowKeyboard = false
    override var canShowKeyboard: Boolean
        get() = _canShowKeyboard
        set(value) {
            _canShowKeyboard = value
            if (!_canShowKeyboard) {
                this.service?.hideKeyboard()
                this.isShowingKeyboard = false
            }
        }

    override fun toggleKeyboard() {
        val service = this.service
        if (service == null || !this.canShowKeyboard) {
            return
        }
        if (this.isShowingKeyboard) {
            service.hideKeyboard()
            this.isShowingKeyboard = false
        } else {
            service.showKeyboard()
            this.isShowingKeyboard = true
        }
    }

    override fun setService(service: Keyboard.Service) {
        this.service = service
    }

    override fun subscribe(client: Keyboard.Client) {
        this.subscriptions.add(client)
    }

    override fun unsubscribe(client: Keyboard.Client) {
        this.subscriptions.remove(client)
    }

    override fun insertText(text: String) {
        if (text.isEmpty()) {
            return
        }
        for (client in this.subscriptions) {
            client.receive(Action.Text(text))
        }
    }

    override fun deleteBackward() {
        for (client in this.subscriptions) {
            client.receive(Action.Backspace)
        }
    }
}